### PR TITLE
10.0 purchase fop shipping

### PR DESCRIPTION
**Pull request by mourad-ehm**
This module allows to define a Free-Of-Paiment (FOP) shipping on supplier.
FOP shipping is a min order amount to got shipping free from supplier.
User can force confirm order if he/she wishes.

### CODE REVIEW

File: purchase_fop_shipping/views/purchase_view.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+					</attribute>
+				</xpath>
+				<field name="origin" position="after">
+					<field name="fop_reached" />
+				</field>
+			</field>
+		</record>
+
+		<record model="ir.ui.view" id="purchase_order_delivery_paid_form_view_inherit">
+			<field name="name">purchase.order.delivery.paid.form.view.inherit</field>
```
Comment (legalsylvain) [inline purchase_fop_shipping/views/purchase_view.xml]: name field is unnessary.

File: purchase_fop_shipping/views/purchase_view.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+	<data>
```
Comment (legalsylvain) [inline purchase_fop_shipping/views/purchase_view.xml]: replace by <odoo> and underindent all the code.

File: purchase_fop_shipping/models/purchase.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+class PurchaseOrder(models.Model):
+
+    _inherit = "purchase.order"
+    fop_reached = fields.Boolean(
+        string='FOP reached',
+        help='Free-Of-Paiment shipping reached',
+        compute='_fop_shipping_reached')
+    force_order_under_fop = fields.Boolean(
+        string='Confirm under FOP',
+        help='Force confirm order under Free-Of-Paiment shipping',)
```
Comment (legalsylvain) [inline purchase_fop_shipping/models/purchase.py]: What about to have a group for this field. Maybe only purchase manager should have the right to check this box ?

File: purchase_fop_shipping/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    'license': 'AGPL-3',
+    'category': 'Purchase',
+    'complexity': 'normal',
+    'depends': ['purchase'],
+    'website': 'http://www.akretion.com/',
+    'data': ['views/purchase_view.xml',
+             'views/partner_view.xml',
+             ],
+    'installable': True,
+    'auto_install': False,
```
Comment (legalsylvain) [inline purchase_fop_shipping/__manifest__.py]: you can remove obsolete or default keys. (complexity, auto_install).

File: purchase_fop_shipping/README.rst [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+   :alt: Odoo Community Association
+   :target: http://odoo-community.org
+
+This module is maintained by the OCA.
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
Comment (legalsylvain) [inline purchase_fop_shipping/README.rst]: could you use the last template ?
https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst
and about == title, you have to have the same number of '=' as the number of char in the tilte.

File: purchase_fop_shipping/README.rst [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+===========================
+
+This module allows to define a Free-Of-Paiment (FOP) shipping on supplier.
+FOP shipping is a min order amount to got shipping free from supplier.
+User can force confirm order if he/she wishes. 
+
+Installation
+============
+
+You need an Odoo/OCB installation v 10.0.
```
Comment (legalsylvain) [inline purchase_fop_shipping/README.rst]: i guess that such lines are not necessary. You can freely remove all the empty (usage) or default section (installation)

File: purchase_fop_shipping/i18n/purchase_fop_shipping.pot [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+#. module: purchase_fop_shipping
+#: model:ir.model,name:purchase_fop_shipping.model_purchase_order
+msgid "Purchase Order"
+msgstr ""
+
+#. module: purchase_fop_shipping
+#: code:addons/purchase_fop_shipping/models/purchase.py:48
+#, python-format
+msgid "You cannot confirm a purchase order with Invoice Control Method 'Based on incoming shipments' that doesn't contain any stockable item."
+msgstr ""
```
Comment (legalsylvain) [inline purchase_fop_shipping/i18n/purchase_fop_shipping.pot]: it seems that your translation file is not up to date;

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Hi @mourad-ehm. 
thanks a lot for this module. I'll use it for sure. Some minor remarks inline.
Otherwise, I think it's a good implementation of the free-of-payment concept.

(code review / no test).

File: purchase_fop_shipping/README.rst [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+Usage
```
Comment (elicoidal) [inline purchase_fop_shipping/README.rst]: remove here as well

File: purchase_fop_shipping/README.rst [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+=====
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/purchase-workflow/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/purchase-workflow/issues/new?body=module:%20purchase_fop_shipping%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (elicoidal) [inline purchase_fop_shipping/README.rst]: adapt to newest template

File: purchase_fop_shipping/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# © 2017 Akretion (http://www.akretion.com)
+#   @author Mourad EL HADJ MIMOUNE <mourad.elhadj.mimoune@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    "name": "Purchase Free-Of-Paiment shipping",
+    'version': '10.0.0.1.0',
```
Comment (elicoidal) [inline purchase_fop_shipping/__manifest__.py]: 10.0.1.0.0

File: purchase_fop_shipping/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    'author': "Akretion,Odoo Community Association (OCA)",
+    'maintainer': 'Akretion',
```
Comment (elicoidal) [inline purchase_fop_shipping/__manifest__.py]: I think you can remove this key

File: purchase_fop_shipping/models/purchase.py (hunk 3 above)
Comment (elicoidal) [reply purchase_fop_shipping/models/purchase.py]: Actually I would create a specific group FOP shipping manager

File: purchase_fop_shipping/models/purchase.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    fop_shipping = fields.Float(
```
Comment (elicoidal) [inline purchase_fop_shipping/models/purchase.py]: shouldnot it be readonly or only for specific group

File: purchase_fop_shipping/models/purchase.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        for record in self:
+            record.fop_reached = record.amount_total >\
+                record.partner_id.fop_shipping
+
+    @api.multi
+    def button_approve(self, force=False):
+        for po in self:
+            if not po.force_order_under_fop and not po.fop_reached:
+                raise UserError(
+                    _('You cannot confirm a purchase order with amount under '
```
Comment (elicoidal) [inline purchase_fop_shipping/models/purchase.py:38]: That looks contradictory for me: in README "FOP shipping is a min order amount to got shipping free from supplier."
We should be able to approve with a fee, no?

File: purchase_fop_shipping/views/purchase_view.xml (hunk 2 above)
Comment (elicoidal) [reply purchase_fop_shipping/views/purchase_view.xml]: replace openerp by odoo

**Review by elicoidal — CHANGES_REQUESTED** [review]
There is some illogical functional behavior in this module.

File: purchase_fop_shipping/models/purchase.py (hunk 3 above)
Comment (AaronHForgeFlow) [reply purchase_fop_shipping/models/purchase.py]: Same remark

File: purchase_fop_shipping/models/purchase.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+from openerp.exceptions import Warning as UserError
+from openerp import models, fields, api, _
+
+
+class PurchaseOrder(models.Model):
+
+    _inherit = "purchase.order"
+    fop_reached = fields.Boolean(
+        string='FOP reached',
+        help='Free-Of-Paiment shipping reached',
```
Comment (AaronHForgeFlow) [inline purchase_fop_shipping/models/purchase.py]: Please correct the typo mistake

Comment (LoisRForgeFlow) [reply purchase_fop_shipping/models/purchase.py]: "Payment" instead of "Paiment".

File: purchase_fop_shipping/models/purchase.py (hunk 3 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/models/purchase.py]: I'm ok for this proposition (group FOP shipping manager)

File: purchase_fop_shipping/__manifest__.py (hunk 4 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/__manifest__.py]: Ok

File: purchase_fop_shipping/README.rst (hunk 6 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/README.rst]: OK

File: purchase_fop_shipping/README.rst (hunk 9 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/README.rst]: Ok

File: purchase_fop_shipping/__manifest__.py (hunk 11 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/__manifest__.py]: You mean 'maintainer' key ?

File: purchase_fop_shipping/models/purchase.py (hunk 12 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/models/purchase.py]: I think readonly is more safe. In some use case supplier can accept to apply exceptionally a low FOP for 1 order. Some user can think that the change is only for the current purchase, but it s not true.

File: purchase_fop_shipping/models/purchase.py (hunk 13 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/models/purchase.py:38]: Yes, if we force validation ( po.force_order_under_fop = true) we approve with a fee.

**Review by legalsylvain — APPROVED** [review]

**Comment by legalsylvain** [comment]
:+1: Thanks !

In general please ping me by name as otherwise I dont see personal responses...

File: purchase_fop_shipping/README.rst [hunk 15]
```diff
@@ -0,0 +1,77 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=================================
+Pruchase Free-Of-Paiment shipping
```
Comment (bealdav) [inline purchase_fop_shipping/README.rst]: s/Pruchase/Purchase

File: purchase_fop_shipping/models/partner.py [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, fields
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    fop_shipping = fields.Float(
+        'FOP shipping', help='Min order amount for FOP shipping',)
```
Comment (bealdav) [inline purchase_fop_shipping/models/partner.py]: order is ambiguous. Here I think it's purchase
s/order/purchase

**Review by bealdav — CHANGES_REQUESTED** [review]

File: purchase_fop_shipping/models/purchase.py (hunk 13 above)
Comment (bealdav) [reply purchase_fop_shipping/models/purchase.py:38]: Mourad added "User can force confirm order if he/she wishes." in readme. So it's OK, no ?

Comment (elicoidal) [reply purchase_fop_shipping/models/purchase.py:38]: clearer! :+1:

**Review by elicoidal — APPROVED** [review]

File: purchase_fop_shipping/README.rst (hunk 15 above)
Comment (mourad-ehm) [reply purchase_fop_shipping/README.rst]: Thanks @bealdav. I fixed typo.

**Comment by bealdav** [comment]
:+1:

**Pull request merged**